osu.Game/Online/OnlineViewContainer.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content which require a local user to be logged in.
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// </summary>
+    public abstract class OnlineViewContainer : Container, IOnlineComponent
+    {
+        private readonly Placeholder placeholder;
+        protected readonly LoadingAnimation LoadingAnimation;
+
+        protected const double TRANSFORM_TIME = 300.0;
+
+        internal readonly Container TransformationTarget;
```

Comment:
why `internal` (if for testing purposes, never expose a property in normal classes for testing purposes).

Reply:
`protected` with exposition in `TestSceneOnlineViewContainer` could work, but then it's kind of not great how you can do things with `TransformationTarget` on your own and *also* define `FadeContent{In,Out}`. Maybe should be `private` instead, and the tests should check the `Box` itself with `ChildrenOfType<>`. Not 100% sure either way.

Reply:
right, `TransformationTarget` should not be visible not even to the sub-class, checking *content* of the `TestOnlineViewContainer` is probably the best, I advocate for that.